metsim/metsim.py

```diff
@@ -179,101 +233,61 @@ def ascii_out_preprocess(self):
-
-        # Create dimensions
-        self.output.createDimension('lat', len(self.lats))
-        self.output.createDimension('lon', len(self.lons))
-        self.output.createDimension('time', n_ts)
-
-        # Create output variables
-        for varname in MetSim.params['out_vars']:
-            self.output.createVariable(varname, 'd', ('time', 'lat', 'lon'))
-
-    def create_io_process(self, queue):
-        """Launch the IO process"""
-        out_preprocess = {"ascii": self.ascii_out_preprocess,
-                          "netcdf": self.netcdf_out_preprocess}
-        out_preprocess[MetSim.params['out_format']]()
-        # Wait for data to come in
-        while 1:
-            data = queue.get()
-            if data == "done":
-                break
-            self.write(data)
-
-        # Close the dataset when done
-        if self.output:
-            self.output.close()
-
-    def write(self, data: dict):
+    def write(self):
         """
         Dispatch to the right function based on the configuration given
         """
         dispatch = {
                 'netcdf': self.write_netcdf,
                 'ascii': self.write_ascii
                 }
-        dispatch[MetSim.params.get('out_format', 'netcdf').lower()](data)
+        dispatch[MetSim.params.get('out_format', 'netcdf').lower()]()
 
-    def write_netcdf(self, data: dict):
+    def write_netcdf(self):
         """Write out as NetCDF to the output file"""
         print("Writing netcdf...")
-        for l in data.keys():
-            lat, lon = l.split("_")
-            for varname in MetSim.params['out_vars']:
-                i, j = self.lat_idx[lat], self.lon_idx[lon]
-                self.output.variables[varname][:, i, j] = (
-                        data[l][varname].values)
-
-    def write_ascii(self, data: dict):
+        self.output.to_netcdf(self.output_filename,
+                              unlimited_dims=['time'],
+                              encoding={'time': {'dtype': 'f8'}})
+
+    def write_ascii(self):
         """Write out as ASCII to the output file"""
         print("Writing ascii...")
-        out_dir = MetSim.params.get('out_dir', './results/')
-        if not os.path.exists(out_dir):
-            os.makedirs(out_dir)
-        for l in data.keys():
-            lat, lon = l.split("_")
-            out_file = '_'.join(["forcing", str(lat), str(lon)])
-            data[l].to_csv(os.path.join(out_dir, out_file), sep='\t')
-
+        shape = self.output.dims
+        for i, j in itertools.product(range(shape['lat']), range(shape['lon'])):
+            if self.output.mask[i, j]>0:
+                lat = self.output.lat.values[i]
+                lon = self.output.lon.values[j]
+                fname = os.path.join(self.params['out_dir'], 
+                            "forcing_{}_{}.csv".format(lat, lon))
+                self.output.isel(lat=i, lon=j).to_dataframe()[
+                        self.params['out_vars']].to_csv(fname)
```

Comment:
you'll save a potential copy if you select the `out_vars` before calling `to_dataframe()`.